contracts/AnchoredView/AnchoredView.sol

```diff
@@ -0,0 +1,278 @@
+            additionalScale = 1e18;
+        }
+
+        // usdcPrice has 30 decimals, so final result has 6
+        return mul(priceInEth, additionalScale) / usdcPrice;
+    }
+
+    /**
+     * @notice Implements the method of the PriceOracle interface of Compound v2 and returns returns the Eth price for an asset.
+     * @dev converts from 1e6 decimals of Open Oracle to 1e(36 - underlyingDecimals) of PriceOracleProxy
+     * @param cToken The cToken address for price retrieval
+     * @return The price for the given cToken address
+     */
+    function getUnderlyingPrice(address cToken) external view returns (uint256) {
+        if (breaker == true) {
+            return readAnchor(cToken);
+        }
+
+        uint256 usdPerToken = _prices[getOracleKey(cToken)];
+
+        if ( usdPerToken == 0 ) {
+            return readAnchor(cToken);
+        } else {
+            uint256 usdPerEth = _prices["ETH"];
+            uint256 ethPerToken = mul(usdPerToken, 1e6) / usdPerEth;
+            uint256 additionalScale = getAdditionalScale(cToken);
+
+            return mul(ethPerToken, additionalScale);
+        }
+    }
+
+    /**
+     * @notice Get the underlying price of a listed cToken asset
+     * @param cTokenAddress The cToken to get the underlying price of
+     * @return The underlying asset price mantissa (scaled by 1e18)
+     */
+    function readAnchor(address cTokenAddress) public view returns (uint) {
+        if (cTokenAddress == cEthAddress) {
+            // ether always worth 1
+            return 1e18;
+        }
+
+        if (cTokenAddress == cUsdcAddress || cTokenAddress == cUsdtAddress) {
+            return anchor.assetPrices(usdcOracleKey);
+        }
+
+        if (cTokenAddress == cDaiAddress) {
+            return anchor.assetPrices(daiOracleKey);
+        }
+
+        if (cTokenAddress == cSaiAddress) {
+            return saiPrice;
+        }
+
+        // otherwise just read from v1 oracle
+        address underlying = CErc20(cTokenAddress).underlying();
+        return anchor.assetPrices(underlying);
+    }
+
+    /// @notice invalidate the source, and fall back to using anchor directly in all cases
+    function invalidate(bytes memory message, bytes memory signature) public {
+        (string memory decoded_message, ) = abi.decode(message, (string, address));
+        require(keccak256(abi.encodePacked(decoded_message)) == keccak256(abi.encodePacked("rotate")), "invalid message must be 'rotate'");
+        require(priceData.source(message, signature) == source, "invalidation message must come from the reporter");
+
+        breaker = true;
+        emit SourceInvalidated(source);
+    }
+
+    /// @notice invalidate the anchor, and fall back to using source without anchor
+
+    /// @dev determine if anchor is stale by checking when usdc was last updated
+    // @dev all anchor prices are converted through usdc price, so if it is stale they are all stale
+    function cutAnchor() public {
```

Comment:
how often this method will be called?

Reply:
it will only do anything once, in the event the current oracle goes down for 24 hours ( or maybe there is a problem here where it would also be possible if ETH / USDC was completely stable for 24 hours )

Reply:
we could also inline this code for "ignoreAnchor" and call it everytime the price is outside of anchor rather than cutting entirely, so that if anchor comes back online, it is used again.